### PR TITLE
Expand effect of dictionary settings in `ColumnFamilyOptions::compression_opts`

### PR DESCRIPTION
**Pull request by ajkr**
In dictionary compression's initial implementation, in order to save CPU overhead, we only enabled it
for bottom level under the assumption that the vast majority of data is
stored there. At that time, there was no
such thing as `ColumnFamilyOptions::bottommost_compression_opts`, so we just
hardcoded disabling dictionary compression in flush and compactions to
non-bottommost level. Now, we have users who generate all their files
through flush and are considering using dictionary compression.

To support such a use case, this PR expands the scope of `ColumnFamilyOptions::compression_opts` to
additionally include flushed files and files generated by compaction to
a non-bottommost level. Users can still get the old behavior by moving
their dictionary settings to `ColumnFamilyOptions::bottommost_compression_opts`
and setting its `enabled` flag.

### CODE REVIEW

**Comment by riversand963** [comment]
Thanks @ajkr for the PR. Just to clarify expected behavior before and after this change.
- Before. RocksDB does not use dictionary compression during flush or compaction to non-bottommost level, even if `ColumnFamilyOptions.compression` is set to dictionary. When `bottommost_compression` is set, `ColumnFamilyOptions.compression` does not affect bottommost compaction. Only when `bottommost_compression` is disabled does `compression` affect bottommost compaction.
- After. RocksDB can use dictionary compression during flush and compaction to non-bottommost levels. If `bottommost_compression` is disabled, `compression` will also affect bottommost compaction. If `bottommost_compression` is enabled, then `bottommost_compression` (not `compression`) will affect bottommost compaction.

**Comment by ajkr** [comment]
> Thanks @ajkr for the PR. Just to clarify expected behavior before and after this change.
> 
> * Before. RocksDB does not use dictionary compression during flush or compaction to non-bottommost level, even if `ColumnFamilyOptions.compression` is set to dictionary. When `bottommost_compression` is set, `ColumnFamilyOptions.compression` does not affect bottommost compaction. Only when `bottommost_compression` is disabled does `compression` affect bottommost compaction.
> * After. RocksDB can use dictionary compression during flush and compaction to non-bottommost levels. If `bottommost_compression` is disabled, `compression` will also affect bottommost compaction. If `bottommost_compression` is enabled, then `bottommost_compression` (not `compression`) will affect bottommost compaction.

Yes, that is exactly right. The motivation is for a use case generates all their files via flush and then just trivial moves to bottom level so (without this PR) they don't have a chance to be compressed with dictionary. I also think the new behavior makes more sense even ignoring any particular use case (we probably would've done it this way initially if `ColumnFamilyOptions::bottommost_compression` had existed at the time).

Will fix the failing test...

**Review by ltamasi — COMMENTED** [review]
Thanks for the patch @ajkr ! LGTM with a few minor questions.

File: HISTORY.md [hunk 1]
```diff
@@ -14,6 +14,9 @@
 ### Public API Change
 * Deprecate `BlockBasedTableOptions::pin_l0_filter_and_index_blocks_in_cache` and `BlockBasedTableOptions::pin_top_level_index_and_filter`. These options still take effect until users migrate to the replacement APIs in `BlockBasedTableOptions::metadata_cache_options`. Migration guidance can be found in the API comments on the deprecated options.
 
+### Behavior Changes
+* The dictionary compression settings specified in `ColumnFamilyOptions::compression_opts` now additionally affect files generated by flush and compaction to non-bottommost level. Previously those settings at most affected files generated by compaction to bottommost level, depending on whether `ColumnFamilyOptions::bottommost_compression_opts` overrode them.
```
[the remark this answers is not in the document]
Comment (ajkr) [reply HISTORY.md]: Done. The migration story in the description and this release note assumed moving dictionary compression settings to `bottommost_compression_opts` is sufficient. Unfortunately `bottommost_compression_opts` is ignored unless `bottommost_compression` is explicitly set - https://github.com/facebook/rocksdb/blob/a388c8cc6b5c5e5edefdc58b856cce396dfad7f7/db/compaction/compaction_picker.cc#L145. I think we should change that before release time. Filed #7631

File: db/db_test2.cc [hunk 2]
```diff
@@ -1390,6 +1391,233 @@ TEST_F(DBTest2, PresetCompressionDictLocality) {
...
+class PresetCompressionDictTest
+    : public DBTestBase,
+      public testing::WithParamInterface<std::tuple<CompressionType, bool>> {
+ public:
+  PresetCompressionDictTest()
+      : DBTestBase("/db_test2", false /* env_do_fsync */) {}
+
+  void SetUp() override {
+    compression_type_ = std::get<0>(GetParam());
+    bottommost_ = std::get<1>(GetParam());
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_test2.cc]: Done.

File: db/db_test2.cc [hunk 3]
```diff
@@ -1390,6 +1391,233 @@ TEST_F(DBTest2, PresetCompressionDictLocality) {
...
+  }
+
+ protected:
+  CompressionType compression_type_;
+  bool bottommost_;
+};
+
+namespace {
+
+bool DictCompressionTypeSupported(CompressionType compression_type) {
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_test2.cc]: Sure, I moved these to the places where their sibling functions live ("util/compression.h" and "options/options_helper.{h,cc}").

File: db/db_test2.cc [hunk 4]
```diff
@@ -1390,6 +1391,233 @@ TEST_F(DBTest2, PresetCompressionDictLocality) {
...
+  options.table_factory.reset(NewBlockBasedTableFactory(bbto));
+  Reopen(options);
+
+  uint64_t prev_compression_dict_misses =
+      TestGetTickerCount(options, BLOCK_CACHE_COMPRESSION_DICT_MISS);
+  Random rnd(301);
+  for (size_t i = 0; i <= kKeysPerFile; ++i) {
+    ASSERT_OK(Put(Key(static_cast<int>(i)), rnd.RandomString(kValueLen)));
+  }
+  dbfull()->TEST_WaitForFlushMemTable();
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_test2.cc]: Done.

File: db/db_test2.cc [hunk 5]
```diff
@@ -1390,6 +1391,233 @@ TEST_F(DBTest2, PresetCompressionDictLocality) {
...
+
+  // If there's a compression dictionary, it should have been loaded when the
+  // flush finished, incurring a cache miss.
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_test2.cc:1445]: Good point. There are lower level ways like checking a `SyncPoint` was called or checking the file's meta-index block has a pointer to a dictionary. But in terms of DB-level APIs (which I tried to use in this test), I do not know of a significantly different way. Maybe we need more properties/stats.

File: db/db_test2.cc [hunk 6]
```diff
@@ -1390,6 +1391,233 @@ TEST_F(DBTest2, PresetCompressionDictLocality) {
...
+#ifndef ROCKSDB_LITE
+  ASSERT_EQ("2,0,1", FilesPerLevel(0));
+#endif  // ROCKSDB_LITE
+
+  uint64_t prev_compression_dict_misses =
+      TestGetTickerCount(options, BLOCK_CACHE_COMPRESSION_DICT_MISS);
+  // This L0->L1 compaction merges the two L0 files into L1. The produced L1
+  // file is not bottommost due to the existing L2 file covering the same key-
+  // range.
+  dbfull()->TEST_CompactRange(0, nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_test2.cc]: Done.

**Review by ajkr — COMMENTED** [review]
Thanks very much for the detailed reviews!

**Review by ltamasi — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**